taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -731,34 +720,48 @@ class KernelGen : public IRVisitor {
     KernelGen *gen;
     std::unique_ptr<ScopedIndent> s;
 
-    ScopedGridStrideLoop(KernelGen *gen) : gen(gen) {
-      // TODO(archibate): what's grid dim actually?
-      size_t stride_size = gen->kernel->program.config.saturating_grid_dim;
-      if (gen->used_tls && stride_size == 0) {
-        // automatically enable grid-stride-loop when TLS used:
-        stride_size = 32;  // seems to be the most optimal number for fem99.py
-      }
-      if (stride_size != 0) {
-        gen->is_grid_stride_loop_ = true;
-        gen->emit("int _sid0 = int(gl_GlobalInvocationID.x) * {};",
-                  stride_size);
-        gen->emit("for (int _sid = _sid0; _sid < _sid0 + {}; _sid++) {{",
-                  stride_size);
-        s = std::make_unique<ScopedIndent>(gen->line_appender_);
-        TI_ASSERT(gen->ps);
-        gen->ps->strides_per_thread = stride_size;
-
-      } else {  // zero regression
-        gen->emit("int _sid = int(gl_GlobalInvocationID.x);");
+    ScopedGridStrideLoop(KernelGen *gen, int const_iterations)
+        : ScopedGridStrideLoop(gen,
+                               fmt::format("{}", const_iterations),
+                               const_iterations) {
+    }
+
+    ScopedGridStrideLoop(KernelGen *gen,
+                         std::string iterations,
+                         int const_iterations = -1)
+        : gen(gen) {
+      gen->emit("int _sid0 = int(gl_GlobalInvocationID.x);");
+      gen->emit("for (int _sid = _sid0; _sid < ({}); _sid += {}) {{",
+                iterations, "int(gl_WorkGroupSize.x * gl_NumWorkGroups.x)");
+      s = std::make_unique<ScopedIndent>(gen->line_appender_);
+
+      TI_ASSERT(gen->ps);
+      if (gen->ps->grid_dim == 0) {
+        // if not specified, guess an optimal grid_dim for different situations
+        // Refs:
+        // https://stackoverflow.com/questions/36374652/compute-shaders-optimal-data-division-on-invocations-threads-and-workgroups
+        if (gen->used_tls) {
+          // TLS reduction
+          gen->ps->grid_dim = std::max(
+              (size_t)const_iterations / gen->ps->block_dim / 32, (size_t)1);
+          gen->ps->block_dim /= 4;
+        } else if (const_iterations > 0) {
+          // const range
+          gen->ps->grid_dim =
+              std::max(((size_t)const_iterations + gen->ps->block_dim - 1) /
+                           gen->ps->block_dim,
+                       (size_t)1);
+        } else {
+          // dynamic range
+          gen->ps->block_dim *= 2;
+          gen->ps->grid_dim = 4;
```

Comment:
Any reason for using `4` here?

Reply:
A rough guess of 1024 threads. What's the best value of grid_dim you used for CUDA?

Reply:
On CUDA we fetch the number of SMs and multiply that number by 16 or 32. This means we usually use a grid_dim of > `500`.

4 or 8 here seems to be too small, and will hardly utilize the SMs.